### ISSUE TITLE
PropertyFilterSpec vs FilterSpec

### ISSUE THREAD

**Issue by tatankat** [vmware/pyvmomi#309]
In pyvmomi, the "PropertyFilterSpec" data object is named "FilterSpec", probably in [QueryTypes.py](/vmware/pyvmomi/blob/39bd020f1a052bf1773a74242c2cfca1837e95e3/pyVmomi/QueryTypes.py#L48).

Is there a reason/cause? Can this be changed/fixed to match the original API? (and how is this file auto-generated?)
If not: the documentation references "PropertyFilterSpec" and should be changed.

**Comment by wiggin15** [comment]
In pyVmomi.vim, there are two ways to use an object: either using the API name, or by using the "full path" defined in the pyVmomi model, e.g.

    >>> from pyVmomi import vim, vmodl
    >>> vim.ServiceContent
    pyVmomi.VmomiSupport.vim.ServiceInstanceContent
    >>> vim.ServiceContent == vim.ServiceInstanceContent
    True
    >>> vim.VirtualDiskDeltaDiskFormat
    pyVmomi.VmomiSupport.vim.vm.device.VirtualDisk.DeltaDiskFormat
    >>> vim.VirtualDiskDeltaDiskFormat == vim.vm.device.VirtualDisk.DeltaDiskFormat
    True

What you're seeing is that the "full path" of PropertyFilterSpec ends with "FilterSpec", but if you try to use the API name, you can only access PropertyFilterSpec like the documented API

    >>> vim.PropertyFilterSpec
    pyVmomi.VmomiSupport.vmodl.query.PropertyCollector.FilterSpec
    >>> vim.PropertyFilterSpec == vmodl.query.PropertyCollector.FilterSpec
    >>> vim.FilterSpec
    AttributeError

In other words, if you use `vim.<attribute>` (recommended) you should always be able to access the pyvmomi classes using their documented API names, but if you try to use the "full path" like vmodl.query.PropertyCollector.FilterSpec, you'll find that the names of many of the objects have a slightly different name - and that's ok.

**Comment by tatankat** [comment]
Thanks for the clear explanation!
